ethereum/spec/src/main/java/tech/pegasys/teku/spec/datastructures/state/beaconstate/BeaconStateInvariants.java

```
@@ -0,0 +1,63 @@
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.ssz.backing.schema.SszPrimitiveSchemas;
+import tech.pegasys.teku.ssz.backing.schema.SszSchema;
+import tech.pegasys.teku.ssz.backing.view.SszPrimitives;
+import tech.pegasys.teku.ssz.sos.SszField;
+
+class BeaconStateInvariants {
+  // Schemas
+  static final SszSchema<SszPrimitives.SszUInt64> GENESIS_TIME_SCHEMA =
+      SszPrimitiveSchemas.UINT64_SCHEMA;
+  static final SszSchema<SszPrimitives.SszBytes32> GENESIS_VALIDATORS_ROOT_SCHEMA =
+      SszPrimitiveSchemas.BYTES32_SCHEMA;
+  static final SszSchema<SszPrimitives.SszUInt64> SLOT_SCHEMA = SszPrimitiveSchemas.UINT64_SCHEMA;
+
+  // Fields
+  static SszField GENESIS_TIME_FIELD =
+      new SszField(0, BeaconStateFields.GENESIS_TIME.name(), GENESIS_TIME_SCHEMA);
+  static SszField GENESIS_VALIDATORS_ROOT_FIELD =
+      new SszField(
+          1, BeaconStateFields.GENESIS_VALIDATORS_ROOT.name(), GENESIS_VALIDATORS_ROOT_SCHEMA);
+  static SszField SLOT_FIELD = new SszField(2, BeaconStateFields.SLOT.name(), SLOT_SCHEMA);
+
+  // Return list of invariant fields
+  public static List<SszField> getInvariantFields() {
+    return List.of(GENESIS_TIME_FIELD, GENESIS_VALIDATORS_ROOT_FIELD, SLOT_FIELD);
+  }
+
+  public static UInt64 extractSlot(final Bytes bytes) {
+    // Check assumptions
+    checkState(GENESIS_TIME_SCHEMA.isFixedSize(), "Expected genesisTime field to be a fixed size");
+    checkState(
+        GENESIS_VALIDATORS_ROOT_SCHEMA.isFixedSize(),
+        "Expected genesisValidatorsRoot field to be a fixed size");
+    checkState(SLOT_SCHEMA.isFixedSize(), "Expected slot field to be a fixed size");
+
+    final int offset =
+        GENESIS_TIME_SCHEMA.getFixedPartSize() + GENESIS_VALIDATORS_ROOT_SCHEMA.getFixedPartSize();
+    final int size = SLOT_SCHEMA.getFixedPartSize();
+
+    // Extract slot data
+    final Bytes slotData = bytes.slice(offset, size);
+    return SLOT_SCHEMA.sszDeserialize(slotData).get();
```

Just thought if we ever might need kind of 'lazy' (and may be lenient) deserialization? Here e.g. you could in theory just deserialize `BeaconState` and call `getSlot()` with almost the same overhead as this custom ssz manipulation. I.e. the parts of an ssz structure would be deserialized on demand

---

Yeah - there's probably a cleaner / more universal way to handle extracting fields.  Just not sure if it's worth putting in the effort for this one use case 🤔?   Also not sure that we'd want to lazy-load everywhere as it seems useful to fail early if we run into a deserialization issue.